src/psyclone/psyir/transformations/intrinsics/array_reduction_base_trans.py

```diff
@@ -0,0 +1,350 @@
+                    raise TransformationError(
+                        f"Unexpected shape for array. Expecting one of "
+                        f"Deferred, Attribute or Bounds but found '{shape}'.")
+
+    # pylint: disable=too-many-locals
+    def apply(self, node, options=None):
+        '''Apply the array-reduction intrinsic conversion transformation to
+        the specified node. This node must be one of these intrinsic
+        operations which is converted to an equivalent loop structure.
+
+        :param node: an array-reduction intrinsic.
+        :type node: :py:class:`psyclone.psyir.nodes.IntrinsicCall`
+        :param options: options for the transformation.
+        :type options: Optional[Dict[str, Any]]
+
+        '''
+        self.validate(node)
+
+        expr, _, mask_ref = self._get_args(node)
+
+        # Step 1: replace all references to arrays within the
+        # intrinsic expressions and mask argument (if it exists) to
+        # array ranges. For example, 'maxval(a+b, mask=mod(c,2.0)==1)'
+        # becomes 'maxval(a(:,:)+b(:,:), mask=mod(c(:,:),2.0)==1)' if
+        # 'a', 'b' and 'c' are 2 dimensional arrays.
+        rhs = expr.copy()
+        _ = UnaryOperation.create(UnaryOperation.Operator.PLUS, rhs)
+        reference2arrayrange = Reference2ArrayRangeTrans()
+        # The reference to rhs becomes invalid in the following
+        # transformation so we keep a copy of the parent here and
+        # reset rhs to rhs_parent.children[0] after the
+        # transformation.
+        rhs_parent = rhs.parent
+        for reference in rhs.walk(Reference):
+            try:
+                reference2arrayrange.apply(reference)
+            except TransformationError:
+                pass
+        # Reset rhs from its parent as the previous transformation
+        # makes the value of rhs become invalid. We know there is only
+        # one child so can safely use children[0].
+        rhs = rhs_parent.children[0]
+        if mask_ref:
+            mask_ref_parent = mask_ref.parent
+            mask_ref_index = mask_ref.position
+            for reference in mask_ref.walk(Reference):
+                try:
+                    reference2arrayrange.apply(reference)
+                except TransformationError:
+                    pass
+            mask_ref = mask_ref_parent.children[mask_ref_index]
+
+        # Step 2: Put the intrinsic's extracted expression (stored in
+        # the 'rhs' variable) on the rhs of an argument with one of
+        # the arrays within the expression being added to the lhs of
+        # the argument. For example if:
+        # x = maxval(a(:,:)+b(:,:))
+        # then
+        # rhs = a(:,:)+b(:,:)
+        # resulting in the following code being created:
+        # a(:,:) = a(:,:)+b(:,:)
+        orig_lhs = node.ancestor(Assignment).lhs.copy()
+        orig_rhs = node.ancestor(Assignment).rhs.copy()
+        array_refs = rhs.walk(ArrayReference)
+        lhs = array_refs[0].copy()
```

Comment:
Add the comment "We need the lhs of the created expression to be a reference from the expression because the ArrayRange2Loop will use it to get the loop bounds"

Reply:
Added, thanks.